From 820362e3254551331177f753844a6a5298f6bcd3 Mon Sep 17 00:00:00 2001
From: Peter Chiu <chui-hao.chiu@mediatek.com>
Date: Thu, 6 Mar 2025 14:26:43 +0800
Subject: [PATCH 166/236] mtk: hostapd: fix aid allocation

When allocating aid to station, do not check the valid links of stations.
Without this patch, two stations may be allocated to the same aid if
they use different links.

Allocate aid starting from 65. Reserve 64 aids for 16 mbss + 3 * 16 mld
affiliated links.

Signed-off-by: Peter Chiu <chui-hao.chiu@mediatek.com>

The maximum length of interface name is 15.
Without this patch, the AP_VLAN name with default config would be
'phy0.0-ap0.sta65' which exceeds the limitation.

Signed-off-by: Peter Chiu <chui-hao.chiu@mediatek.com>

Free aid bitmap when station is removed.
We add a offset when we get aid from bitmap so we also need to cut
the offset when clean bitmap.

Signed-off-by: Peter Chiu <chui-hao.chiu@mediatek.com>

Fix non-tx bss should use tx hapd aid pool to free sta aid.

Signed-off-by: Allen Ye <allen.ye@mediatek.com>
Signed-off-by: Peter Chiu <chui-hao.chiu@mediatek.com>
---
 src/ap/ap_drv_ops.c |  4 ++--
 src/ap/ieee802_11.c | 16 ++++++----------
 src/ap/sta_info.c   |  8 +++++---
 src/ap/ucode.c      |  4 ++--
 4 files changed, 15 insertions(+), 17 deletions(-)

diff --git a/src/ap/ap_drv_ops.c b/src/ap/ap_drv_ops.c
index 0ad836563..5b21a95ea 100644
--- a/src/ap/ap_drv_ops.c
+++ b/src/ap/ap_drv_ops.c
@@ -410,13 +410,13 @@ int hostapd_set_wds_sta(struct hostapd_data *hapd, char *ifname_wds,
 	if (hapd->conf->apup && hapd->conf->apup_peer_ifname_prefix[0]) {
 		mRet = os_snprintf(
 		            ifName, sizeof(ifName), "%s%d",
-		            hapd->conf->apup_peer_ifname_prefix, aid);
+		            hapd->conf->apup_peer_ifname_prefix, aid - 64);
 	}
 	else
 #endif // def CONFIG_APUP
 		mRet = os_snprintf(
 		            ifName, sizeof(ifName), "%s.sta%d",
-		            hapd->conf->iface, aid);
+		            hapd->conf->iface, aid - 64);
 
 	if (mRet >= (int) sizeof(ifName))
 		wpa_printf(MSG_WARNING,
diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index fe38d6b96..d6db34e94 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -3704,22 +3704,15 @@ static u32 hostapd_get_aid_word(struct hostapd_data *hapd,
 
 	/* Do not assign an AID that is in use on any of the affiliated links
 	 * when finding an AID for a non-AP MLD. */
-	if (hapd->conf->mld_ap && sta->mld_info.mld_sta) {
+	if (hapd->conf->mld_ap) {
 		int j;
 
 		for (j = 0; j < MAX_NUM_MLD_LINKS; j++) {
 			struct hostapd_data *link_bss;
 
-			if (!sta->mld_info.links[j].valid)
-				continue;
-
 			link_bss = hostapd_mld_get_link_bss(hapd, j);
-			if (!link_bss) {
-				/* This shouldn't happen, just skip */
-				wpa_printf(MSG_ERROR,
-					   "MLD: Failed to get link BSS for AID");
+			if (!link_bss)
 				continue;
-			}
 
 			aid_word |= link_bss->sta_aid[i];
 		}
@@ -3764,7 +3757,10 @@ int hostapd_get_aid(struct hostapd_data *hapd, struct sta_info *sta)
 	}
 	if (j == 32)
 		return -1;
-	aid = i * 32 + j + (1 << hostapd_max_bssid_indicator(hapd));
+
+	/* Allocate aid starting from 65 */
+	/* Reserve 16 mbss + 3 * 16 mld affiliated links */
+	aid = i * 32 + j + 65;
 	if (aid > 2007)
 		return -1;
 
diff --git a/src/ap/sta_info.c b/src/ap/sta_info.c
index d19d72d5f..6c4a5c1d9 100644
--- a/src/ap/sta_info.c
+++ b/src/ap/sta_info.c
@@ -275,9 +275,11 @@ void ap_free_sta(struct hostapd_data *hapd, struct sta_info *sta)
 	ap_sta_hash_del(hapd, sta);
 	ap_sta_list_del(hapd, sta);
 
-	if (sta->aid > 0)
-		hapd->sta_aid[(sta->aid - 1) / 32] &=
-			~BIT((sta->aid - 1) % 32);
+	if (sta->aid - 64 > 0) {
+		struct hostapd_data *tx_bss = hostapd_mbssid_get_tx_bss(hapd);
+		tx_bss->sta_aid[(sta->aid - 1 - 64) / 32] &=
+			~BIT((sta->aid - 1 - 64) % 32);
+	}
 
 	hapd->num_sta--;
 	if (sta->nonerp_set) {
diff --git a/src/ap/ucode.c b/src/ap/ucode.c
index 161e521e1..2cbdec3dd 100644
--- a/src/ap/ucode.c
+++ b/src/ap/ucode.c
@@ -849,8 +849,8 @@ uc_hostapd_bss_rename(uc_vm_t *vm, size_t nargs)
 		if (!(sta->flags & WLAN_STA_WDS) || sta->pending_wds_enable)
 			continue;
 
-		snprintf(cur_name, sizeof(cur_name), "%s.sta%d", prev_ifname, sta->aid);
-		snprintf(new_name, sizeof(new_name), "%s.sta%d", ifname, sta->aid);
+		snprintf(cur_name, sizeof(cur_name), "%s.sta%d", prev_ifname, sta->aid - 64);
+		snprintf(new_name, sizeof(new_name), "%s.sta%d", ifname, sta->aid - 64);
 		hostapd_drv_if_rename(hapd, WPA_IF_AP_VLAN, cur_name, new_name);
 	}
 
-- 
2.45.2

